java/src/test/java/ai/onnxruntime/OnnxTensorTest.java

```diff
@@ -87,6 +87,83 @@ public void testScalarCreation() throws OrtException {
     }
   }
 
+  @Test
+  public void testBufferCreation() throws OrtException {
+    OrtEnvironment env = OrtEnvironment.getEnvironment();
+
+    float[] arrValues = new float[] {0, 1, 2, 3, 4};
+    try (OnnxTensor t = OnnxTensor.createTensor(env, arrValues)) {
+      // array creation isn't backed by buffers
+      Assertions.assertFalse(t.isCopy());
+      Assertions.assertFalse(t.getBufferRef().isPresent());
+      FloatBuffer buf = t.getFloatBuffer();
+      float[] output = new float[arrValues.length];
+      buf.get(output);
+      Assertions.assertArrayEquals(arrValues, output);
+
+      // Can't modify the tensor through this buffer.
+      buf.put(0, 25);
+      Assertions.assertArrayEquals(arrValues, output);
+    }
+
+    FloatBuffer nonDirectBuffer = FloatBuffer.allocate(5);
+    nonDirectBuffer.put(arrValues);
+    nonDirectBuffer.rewind();
+    try (OnnxTensor t = OnnxTensor.createTensor(env, nonDirectBuffer, new long[] {1, 5})) {
+      // non-direct buffers trigger a copy
+      Assertions.assertTrue(t.isCopy());
+      // tensors backed by buffers can get the buffer ref back out
+      Assertions.assertTrue(t.getBufferRef().isPresent());
+      FloatBuffer buf = t.getFloatBuffer();
+      float[] output = new float[arrValues.length];
+      buf.get(output);
+      Assertions.assertArrayEquals(arrValues, output);
+
+      // Can't modify the tensor through getFloatBuffer.
+      buf.put(0, 25);
+      Assertions.assertArrayEquals(arrValues, output);
+
+      // Can modify the tensor through getBufferRef.
+      FloatBuffer ref = (FloatBuffer) t.getBufferRef().get();
+      ref.put(0, 25);
+      buf = t.getFloatBuffer();
+      buf.get(output);
+      Assertions.assertEquals(25, output[0]);
+    }
+
+    FloatBuffer directBuffer =
```

Comment:
Some comments would be helpful. To me, the difference between this test case and the previous one is not apparent.

Reply:
This case tests the behaviour when using a direct byte buffer (which is a byte buffer wrapped around a native pointer allocated by malloc on the native heap, but with the cleanup dealt with by the JDK). The code above this tests using a non-direct byte buffer (which is a byte buffer wrapped around a `byte[]` allocated on the Java heap). It's unfortunate that this distinction isn't visible in the type system as that would make it easier to show what's going on. I can put some more comments in.

Reply:
The confusion on my part remains. Here and the above there is a `FloatBuffer`. But in one case there is a copy, and in this case there is not a copy. How does the user understands?

Reply:
We need non-direct because I can't make it clear in the type system as both kinds of `FloatBuffer` live in the same public type. The direct & heap subclasses of the `java.nio.Buffer` implementations are package-private  (https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template). Then the choice becomes either throwing a runtime exception if the user passes in a non-direct buffer, or copying. I chose copying, and it's documented that you need to supply a direct buffer to get the no copy fast path.